pkgtools/pkgchk/files/pkgchk.8
.\"	$NetBSD: pkgchk.8,v 1.3 2001/07/07 12:17:35 abs Exp $
.\"
.\" Copyright (c) 2001 by David Brownlee (abs@netbsd.org)
.\" Absolutely no warranty.
.\"
.Dd June 28, 2001
.Dt PKGCHK 1
.Sh NAME
.Nm pkgchk
.Nd check, and optionally update, installed packages
.Sh SYNOPSIS
.Nm
.Op Fl abhinsv
.Op Fl D Ar tags
.Op Fl U Ar tags
.Sh DESCRIPTION
.Nm
verifies that the versions of installed packages matches those in
pkgsrc, optionally adding missing or updating mismatched packages.
.Pp
It can verify all installed packages, or check against a set of packages
based on the hostname, architecture, and other characteristics.
.Pp
When updating packages that depend on each other
.Nm
will skip dependent packages to reduce unnecessary rebuilding.
.Pp
.Sy Options
.Bl -tag -width xxxxxxxx
.It Fl D Ar tags
Add the comma separated list of tags to those checked when parsing
pkgchk.conf.
.It Fl U Ar tags
Remove the comma separated list of tags from those checked when
parsing pkgchk.conf.
.It Fl a
Automatically add any missing packages. If a binary package exists
in
.Ev ${PACKAGES}
it will be used in preference to building from source, unless the
.Fl b
or
.Fl s
options are given.
.It Fl b
Limit additions or updates to using binary packages. If
DEPENDS_TARGET=package
is set in
.Pa /etc/mk.conf ,
then
.Nm
will automatically generate binary packages for later reuse.
.It Fl h
Brief help.
.It Fl i
Check versions of installed packages, ignoring pkgchk.conf.
.It Fl n
Display actions that would be taken, but do not change anything.
.It Fl s
Limit additions or updates to building from source.
.It Fl u
Automatically update any missing packages. If a binary package exists
in PACKAGES it will be used in preference to building from source, unless the
.Fl b
or
.Fl s
options are given.
.Pp
Note: If the update fails (particularly when building from source), the system
will be left with missing packaged.
.It Fl v
Verbose - list the tags set, and all packages checked.
.El
.Sh FILE FORMAT
.Pp
Each non comment line in pkgsrc.conf should contain a package directory
(such as sysutils/skill) followed by a list of tags. If a tag starts with
a hyphen (-), then any machine matching will be excluded, otherwise if any
tags match the current machine the package will be included. The tag
.Em *
matches all machines.
.Pp
The default set of tags used to determine which packages to match in
pkgsrc.conf are equivalent to the output of the following with any spaces
converted to hypens (-): hostname -s, hostname, uname -srm, uname -sr, uname -sm, uname -s, uname -r, uname -m. If 
.Pa /usr/X11R6/lib/libX11.so
or
.Pa /usr/X11R6/lib/libX11.a
is present,
.Em x11
is added to the list of tags.
To see tags for the current machine, run
.Nm
with the
.Fl v
option.
.Sh ENVIRONMENT
.Nm
uses the following environment variables.
.Bl -tag -width xxxx
.It Ev PKGSRCDIR
Base of pkgsrc tree. If not set in environment then read from
.Pa /etc/mk.conf .
Defaults to
.Pa /usr/pkgsrc
.It Ev PACKAGES
Location of binary packages. If not set in environment then read from
.Pa /etc/mk.conf .
Defaults to
.Pa ${PKGSRCDIR}/packages.
.Pp
Unless in a completely homogeneous environment (every machine running
exactly the same OS version and architecture) setting
.Ev ${PACKAGES}
in
.Pa /etc/mk.conf
to a value such as
.Bd -literal
${PKGSRCDIR}/packages/${LOWER_OPSYS}-${OS_VERSION}-${MACHINE_ARCH}

.Ed
is strongly recommended. If 
.Em cpuflags
(devel/cpuflags) is being used to optimally target individual CPU types,
then
.Ev ${CPU_DIR}
should be appended to 
.Ev ${PACKAGES} .
.It Ev PKGCHK_CONF
Pathname to pkgchk configuration file. If not set in environment
then read from
.Pa /etc/mk.conf .
Defaults to
.Pa ${PKGSRCDIR}/pkgchk.conf .
.El
.Sh AUTHOR
David Brownlee <abs@netbsd.org>
.Sh CAVEATS
Updating packages on which other packages depend can currently only be done
via source (not binary packages).
.Pp
If both package
.Em a
and package
.Em b
are to be updated, and
.Em a
depends on
.Em b ,
.Nm
will correctly skip the update of
.Em a .
However, if 
.Em a
depends on
.Em b
and
.Em c ,
and all three are marked for update,
.Nm
will update
.Em b
and
.Em c
in two separate passes, resulting in unnecessary rebuilding of
.Em a
(and potentially other packages).


doc/NEWS
# $NetBSD: NEWS,v 1.33 2025/12/03 21:57:26 leot Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

m68k softfloat (m68ksf) support [nat 2025-10-22]

        Support for m68k softfloat was added to pkgsrc.

Default versions bumped [wiz 2025-10-06]

	PYTHON_VERSION_DEFAULT is now 313.

Default MySQL implementation changed to MariaDB 11.4 [nia 2024-12-25]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.11 to MariaDB 11.4.

Default versions bumped [wiz 2024-08-12]

	PYTHON_VERSION_DEFAULT is now 312.

Default versions bumped [taca 2024-02-10]

	RUBY_VERSION_DEFAULT is now 32.

Default MySQL implementation changed to MariaDB 10.11 [nia 2024-01-23]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.6 to MariaDB 10.11.

Default versions bumped [wiz 2023-08-14]

	PYTHON_VERSION_DEFAULT is now 311.

Support for USE_CC_FEATURES and USE_CXX_FEAUTRES [nia 2023-07-18]

	USE_CC_FEATURES and USE_CXX_FEATURES can now be used
	to specify a minimum level of C and C++ support required
	to compile a package, including specific libstdc++ features.

	ALLOW_NEWER_COMPILER can be set to no in mk.conf to skip
	building packages that require a newer compiler than the system
	provides.

Default versions bumped [nia 2022-06-30]

	PYTHON_VERSION_DEFAULT is now 310.
	PGSQL_VERSION_DEFAULT is now 14.
	LUA_VERSION_DEFAULT is now 54.

Dropped support for compiling glib2-using packages without introspection
support [wiz 2022-06-07]

PKGSRC_USE_MKTOOLS support added [jperkin 2022-06-07]

        Added C-based tools to speed up various parts of the
        mk infrastructure via PKGSRC_USE_MKTOOLS=yes in mk.conf.

        For packages with large numbers of dependencies, you should see
        anywhere up to around 10x speed-up for the 'wrapper' phase, and
        perhaps more importantly a huge reduction in system time usage.

Linux infrastructure adapted to non-standard distributions [nia 2022-05-01]

        The Linux pkgsrc infrastructure was adapted to support Linux
        distributions where binaries and shared objects are not
        stored in a normal Unix-like file system layout.  It's hoped
        that this will be useful for pkgsrc development on platforms
        like NixOS and Gobo Linux.

UnixWare support [nia 2021-12-26]

        Support for SCO/Xinuos UnixWare was improved to the point where
        it is possible to complete a bootstrap on this platform.

SHA1 distfile hashes deprecated [nia 2021-10-07]

        Distfile hash algorithms have been changed, keeping SHA512,
        adding BLAKE2s, and removing SHA1 and RMD160.  New packages
        and updates will only generate SHA512 and BLAKE2s, and the
        stored hashes have been adjusted.

Default Python version changed to 3.9 [nia 2021-10-01]

Increased enabled-by-default hardening options [nia 2021-09-27]

	PKGSRC_USE_MKPIE was enabled on some architectures.
	Packages that provide static libraries to other packages
	should be rebuilt with MKPIE enabled.
	Some packages may experience runtime errors on platforms
	like i386 - usually this is due to unsafe assembly code,
	which should be disabled when MKPIE is enabled.

	PKGSRC_USE_SSP was bumped from "yes" to "strong".  This
	is not expected to have a noticable effect.

	More information:
	https://www.NetBSD.org/docs/pkgsrc/hardening.html
	(or see "Security hardening" section of doc/pkgsrc.txt)

Default MySQL implementation changed to MariaDB 10.6 [nia 2021-09-27]

Default PostgreSQL changed to 13 [nia 2021-06-28]

Default MySQL implementation changed to MariaDB [nia 2021-06-24]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MySQL 5.7 to MariaDB 10.5.

	This is configurable by setting MYSQL_VERSION_DEFAULT in mk.conf.

	PHP packages are no longer tied to a specific MySQL version
	due to the native driver being enabled by default.

Python 3.8 default [nia 2020-12-04]
	The default Python version was switched to 3.8.

Default database location changed for NetBSD [wiz 2020-12-02]
	The database of installed packages now defaults to ${PREFIX}/pkgdb
	instead of /var/db/pkg on NetBSD. This matches the default on
	other operating systems and cases where a non-default bootstrap
	is used.

	To deal with this change, see this blog post:

	https://pkgsrc.org/news/pkgdb-change/

	Since this change affects the packaging tools only, there is no change
	to binary packages and they do not need to be rebuilt.

lua 5.3 default [nia 2020-03-30]
	The default Lua version was switched to 5.3.
	Since many packages specify a preferred Lua version, this is mainly
	expected to affect users of Lua modules who were using the default
	settings.

gfortran default [nia 2020-03-30]
	gfortran is now the default Fortran compiler instead of g95.
	gfortran is expected to support more pkgsrc platforms and be
	better at compiling newer Fortran code.

python 3.7 default [maya 2019-04-24]
	As Python 2.x versions are nearing end of life, we have switched
	the default to a newer version.
	Python 2.7 packages are still available, and most python packages
	are built for all available python versions.

Lua multiversion support [alnsn 2014-05-03]
	Lua packages lang/lua51 and lang/lua52 have been modified
	to coexist without conflicts, infrastructure for supporting
	multiple versions of Lua has been put in place and all
	dependent packages have been adapted to the new infrastructure.

SMF support is introduced [jperkin 2014-03-11]

Unprivileged staged installation is default [asau 2012-08-14]
	pkgsrc has changed to use unprivileged staged installation by default.
	This makes it possible to prevent some installation mistakes.
	It also makes detection of package conflicts easier.
	Previously you had to set PKG_DESTDIR_SUPPORT to "user-destdir"

pkgtools/pkgchk/files/pkgchk.sh
#!/bin/sh -e
#
# $Id: pkgchk.sh,v 1.4 2001/07/07 12:17:35 abs Exp $
#
# TODO: Handle and as well as or tags (eg: i386+x11)
# TODO: Order updates based on DEPENDENCIES.
# TODO: Handle updates with dependencies via binary packages

extract_variables()
    {
    # Establish PKGSRCDIR
    #

    if [ -z "$PKGSRCDIR" ];then
	if [ -f /etc/mk.conf ] ;then
	    eval `printf 'all:\n\t@echo PKGSRCDIR=${PKGSRCDIR}\n' | make -f - -f /etc/mk.conf`
	fi
	if [ -z "$PKGSRCDIR" ];then
	    PKGSRCDIR=/usr/pkgsrc
	fi
    fi
    if [ ! -d $PKGSRCDIR ];then
	echo "Unable to locate PKGSRCDIR '$PKGSRCDIR'"
	exit 1;
    fi

    # Now we have PKGSRCDIR, use it to determine PACKAGES, and PKGCHK_CONF
    #
    eval `printf 'CATEGORIES=x\nDISTNAME=x\nall:\n
	@echo PACKAGES=${PACKAGES}
	@echo PKGCHK_CONF=${PKGCHK_CONF}
    ' | (cd $PKGSRCDIR ; make -f - -f $PKGSRCDIR/mk/bsd.prefs.mk)`

    if [ -z "$PACKAGES" ];then
	PACKAGES=$PKGSRCDIR/packages
    fi

    if [ -z "$PKGCHK_CONF" ];then
	PKGCHK_CONF=$PKGSRCDIR/pkgchk.conf
    fi
    }

pkg_install()
    {
    PKGNAME=$1
    PKGDIR=$2
    INSTALL=$3
    if [ -n "$opt_b" -a -f $PACKAGES/All/$PKGNAME.tgz ] ; then
	if [ $INSTALL = U ];then
	    echo "pkg_delete $PKGNAME"
	    if [ -z "$opt_n" ];then
		if pkg_delete $PKGNAME ; then
		    echo "Deleted $PKGNAME"
		else
		    echo "Can only update packages with dependencies via -s"
		    exit 1
		fi
	    fi
	fi
	echo "pkg_add $PKGNAME.tgz"
	if [ -z "$opt_n" ];then
	    pkg_add $PACKAGES/All/$PKGNAME.tgz
	fi
    elif [ -n "$opt_s" ]; then
	echo "make update for $PKGNAME"
	cd $PKGSRCDIR/$PKGDIR
	if [ -z "$opt_n" ];then
	    make update
	fi
    fi
    }

args=`getopt D:IU:abchinsuv $*`
if [ $? != 0 ]; then
    opt_h=1
fi
set -- $args
while [ $# != 0 ]; do
    case "$1" in
	-D )	opt_D="$2" ; shift;;
	-U )	opt_U="$2" ; shift;;
	-a )	opt_a=1 ; opt_c=1 ;;
	-b )	opt_b=1 ;;
	-c )	opt_c=1 ;;
	-h )	opt_h=1 ;;
	-i )	opt_i=1 ;;
	-n )	opt_n=1 ;;
	-s )	opt_s=1 ;;
	-u )	opt_u=1 ;;
	-v )	opt_v=1 ;;
	-- )	break ;;
    esac
    shift
done

if [ -z "$opt_b" -a -z "$opt_s" ];then
    opt_b=1; opt_s=1;
fi

if [ -z "$opt_a" -a -z "$opt_c" -a -z "$opt_i" ];then
    echo "Must specify at least one of -a, -c, or -i";
    echo
    opt_h=1;
fi

if [ -n "$opt_h" -o $# != 1 ];then
    echo 'Usage: pkgchk [opts]
	-D tags Comma separated list of additional pkgchk.conf tags to set
	-U tags Comma separated list of pkgchk.conf tags to unset
	-a      Add all missing packages (implies -c)
	-b      Limit installations to binary packages
	-c      Check installed packages against pkgchk.conf
	-h      This help
	-i	Check versions of installed packages (ignore pkgchk.conf)
	-n	Display actions that would be taken, but do not perform them
	-s      Limit installations to building from source
	-u      Update all mismatched packages
	-v      Verbose

pkgchk verifies installed packages against pkgsrc.
The most common usage is 'pkgchk -i' to check all installed packages.
For more advanced usage, including defining a set of desired packages based
on hostname and type, see pkgchk(8).
'
    exit 1
fi

extract_variables

if [ -n "$opt_i" ];then
    PKGDIRLIST=`pkg_info -B \* | awk '/PKGPATH= /{print $2}'`
fi

if [ -n "$opt_c" ];then

    if [ ! -r $PKGCHK_CONF ];then
	echo "Unable to read PKGCHK_CONF '$PKGCHK_CONF'"
	exit 1;
    fi

    # Determine list of tags
    #
    TAGS="`hostname -s`,`hostname`,`uname -srm | awk '{print $1"-"$2"-"$3","$1"-"$2","$1"-"$3","$1","$2","$3}'`"
    if [ -f /usr/X11R6/lib/libX11.so -o /usr/X11R6/lib/libX11.a ];then
	TAGS="$TAGS,x11"
    fi
    if [ -n "$opt_D" ];then
	TAGS="$TAGS,$opt_D"
    fi
    if [ -n "$opt_v" ];then
	echo "set   TAGS=$TAGS"
	echo "unset TAGS=$opt_U"
    fi

    # Extract list of valid pkgdirs
    #
    PKGDIRLIST="$PKGDIRLIST "`awk -v setlist=$TAGS -v unsetlist=$opt_U '
    BEGIN {
	split(setlist, tmp, ",");
	for (tag in tmp)
	    { taglist[tmp[tag]] = 1; }
	taglist["*"] = 1;
	split(unsetlist, tmp, ",");
	for (tag in tmp)
	    {
	    print "UN "notaglist[tag] > "/dev/stderr"
	    delete taglist[tmp[tag]]
	    }
	taglist["*"] = "*"
   	}
    {
    sub("#.*", "");
    split($0, chklist, "[ 	]*");
    need = 0;
    for (chk in chklist)			# For each word on the line
	{
	if (sub("^-", "", chklist[chk]))	# If it begins with a '-'
	    {
	    if (chklist[chk] in taglist)	# If match, discard
		{ next; }
	    }
	else
	    {
	    if (chklist[chk] in taglist)	# If match, note needed
		{ need = 1; }
	    }
	}
    if (need)
	{ print $1 }
    }
    ' < $PKGCHK_CONF
    `
fi

# Check packages are installed
#
for pkgdir in $PKGDIRLIST ; do

    cd $PKGSRCDIR/$pkgdir
    # Use 'make x' rather than 'make all' to avoid potential licence errors
    pkgname=`printf 'x:\n\t@echo ${PKGNAME}\n'|make -f - -f Makefile x` || true
    if [ -z "$pkgname" ]; then
	echo "Unable to extract PKGNAME for $pkgdir"
	exit 1
    fi
    if [ ! -d /var/db/pkg/$pkgname ];then
	echo -n "$pkgname: "
	pkg=`echo $pkgname | sed 's/-[0-9].*//'`
	pkginstalled=`pkg_info -e $pkg || true`
	INSTALL=
	if [ -n "$pkginstalled" ];then
	    echo -n "version mismatch - $pkginstalled"
	    if [ -n "$opt_u" ]; then
		DO_UPDATE="$DO_UPDATE $pkgname $pkgdir"
	    fi
	else
	    echo -n "missing"
	    if [ -n "$opt_a" ] ; then
		DO_INSTALL="$DO_INSTALL $pkgname $pkgdir"
	    fi
	fi
	if [ -f $PACKAGES/All/$pkgname.tgz ] ;then
	    echo -n " (binary package available)"
	fi
	echo
    elif [ -n "$opt_v" ];then
	echo "$pkgname: OK"
    fi
done

if [ -n "$DO_UPDATE" ];then
    echo "Updating..."

    # Generate list including packages which depend on updates
    #
    set -- $DO_UPDATE
    while [ $# != 0 ]; do
	PKGNAME=`echo $1 | sed 's/-[0-9].*//'`
	if [ -f /var/db/pkg/$PKGNAME-[0-9]*/+REQUIRED_BY ];then
	    LIST="$LIST$1|$2|`cat /var/db/pkg/$PKGNAME-[0-9]*/+REQUIRED_BY`\n"
	else
	    LIST="$LIST$1|$2\n"
	fi
	shift ; shift;
    done

    # drop any packages whose 'parents' are also to be updated
    #
    DO_UPDATE=`printf "$LIST" | awk -F '|' '
    {
    pkg2dir[$1] = $2
    split($3, deplist, " ")
    for (pkg in deplist)
	{
	dep = deplist[pkg]
	sub("-[0-9].*", "", dep) # Strip version
	covered[dep] = 1
	}
    }
    END {
    for (pkg in pkg2dir)
	{
	chk = pkg
	sub("-[0-9].*", "", chk); # Strip version
	if (!covered[chk])
	    print pkg" "pkg2dir[pkg]
	}
    }
    '`

    set -- $DO_UPDATE
    while [ $# != 0 ]; do
	pkg_install $1 $2 U
	shift ; shift;
    done
fi

if [ -n "$DO_INSTALL" ];then
    echo "Installing..."
    set -- $DO_INSTALL
    while [ $# != 0 ]; do
	pkg_install $1 $2 I
	shift ; shift;
    done
fi
